### DISCUSSION TITLE
How to run SqlAlchemy test suite with Oracle's new python driver python-oracledb in Thick mode

### DISCUSSION THREAD

**Discussion by Kajalagg98** [sqlalchemy/sqlalchemy#8147] [Usage Questions] [answered]
Thanks for adding the support for Oracle's New Python driver module 'oracledb'.

Since by default the driver runs in 'Thin' mode, so following command works fine for running the SQLAlchemy test suite with oracledb's Thin mode.
`pytest --dburi=oracle+oracledb://user:pswd@host:port/service_name test/`
But for running the test suite with Thick mode, I need to explicitly make a call to init_oracle_client(). Any other alternative to run the test suite with Thick mode such that it does not require calling init_oracle_client() explicitly from my end? Something that I can add in above command to allow the test suite to run with oracledb's Thick mode?

**Comment by CaselIT** [comment, marked as answer]
Hi,

I don't think there is a way at the moment, since "thick mode" cannot be turned on with an url option, but only by passing an argument to `create_engine`.
Would patching the dialect file of oracledb work for you? This patch will enable thick mode by default for the oracle db:
```diff
diff --git a/lib/sqlalchemy/dialects/oracle/oracledb.py b/lib/sqlalchemy/dialects/oracle/oracledb.py
index bbe801cd4..8437fbc5a 100644
--- a/lib/sqlalchemy/dialects/oracle/oracledb.py
+++ b/lib/sqlalchemy/dialects/oracle/oracledb.py
@@ -76,7 +76,7 @@ class OracleDialect_oracledb(_OracleDialect_cx_oracle):
             **kwargs,
         )

-        if thick_mode is not None:
+        if thick_mode is not None or True:
             kw = thick_mode if isinstance(thick_mode, dict) else {}
             self.dbapi.init_oracle_client(**kw)
```

It that would not work we could setup an option to add to the testing url only to enable it


**EDIT: this is now outdated. pass `--oracle-thick-mode` when calling pytest. See [this comment](https://github.com/sqlalchemy/sqlalchemy/discussions/8147#discussioncomment-2988195)**

**Reply by Kajalagg98** [reply]
With this patch I can see many SQLAlchemy tests failing with the following error:
`AttributeError: 'NoneType' object has no attribute 'init_oracle_client'`

I think it would be better if some option is added to the testing url to enable the "thick mode".

**Reply by CaselIT** [reply]
I cannot reproduce that error. Also it doesn't seem connected to the patch.
In what test was it raised? Do you have a full stack trace?

**Reply by Kajalagg98** [reply]
Like these files inside sqlalchemy/test/dialect/oracle/ : test_compiler.py, test_dialect.py and test_types.py.
For stack trace, consider this test case: 
```
DialectTypesTest.test_type_adapt_nchar[sqlalchemy.dialects.oracle.oracledb-start0-String]
Traceback (most recent call last):
  File "/Users/kajalaggarwal/oracle_project/sqlalchemy/test/dialect/oracle/test_types.py", line 125, in test_type_adapt_nchar
    dialect = module.dialect(use_nchar_for_unicode=True)
  File "/Users/kajalaggarwal/oracle_project/sqlalchemy/lib/sqlalchemy/dialects/oracle/oracledb.py", line 81, in __init__
    self.dbapi.init_oracle_client(**kw)
AttributeError: 'NoneType' object has no attribute 'init_oracle_client'
```

Instead of using the above patch, if I explicitly make a call to init_oracle_client() inside lib/sqlalchemy/testing/plugin/pytestplugin.py, then I don't see this AttributeError.

**Reply by CaselIT** [reply]
Thanks, I'll try taking a look.

@zzzeek in case we want to try with a test only query string parameter I would have to support in this function `update_db_opts` in provisioning right?

**Reply by zzzeek** [reply]
not sure if I like query string param here because thick_mode has all those other options you can pass in also, but additionally, having it in the URL implies we can pass ``--dburl`` twice with the two kinds of engines, but that wont actually work since it's a global setting.

how about straight up add a whole new option ``--oracle-thick-mode``  to plugin_base?  we have dialect-specific options here like ``--postgresql-template-db`` already and this is a very special case.   you then consume it in oracle provisioning inside the `update_db_opts` as you noted.

**Reply by CaselIT** [reply]
That was the original reason why I didn't add that to the url.
I can add the option to pytest

**Reply by CaselIT** [reply]
> Like these files inside sqlalchemy/test/dialect/oracle/ : test_compiler.py, test_dialect.py and test_types.py.

Ok I could reproduce these error and I've fixed it.
In https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3937 I've added the option `--oracle-thick-mode` to enable testing with thick mode by calling pytest like this `pytest --dburi oracle+oracledb://the-url  --oracledb-thick-mode test/dialect/oracle/`.

Could you test if it's working for you?

**Reply by Kajalagg98** [reply]
Is this patch available on main branch or on some separate branch? How should I test it?

**Reply by CaselIT** [reply]
It's not yet merged. You can download a it as a diff from the gerrit page. This should be a direct link https://gerrit.sqlalchemy.org/changes/sqlalchemy%2Fsqlalchemy~3937/revisions/1/patch?zip

**Reply by Kajalagg98** [reply]
I have tested with this patch and its working all GOOD!
Let me know once this is merged to main branch.
Thanks for this quick support!!

**Reply by CaselIT** [reply]
Hey @Kajalagg98 it's merged in the main branch

**Reply by Kajalagg98** [reply]
Thanks @CaselIT !
It would be helpful if something related to this will be added to README.unittests.rst. Thoughts?

**Reply by CaselIT** [reply]
sure, feel free to send a PR if you want to add some changes in that file in the oracle section

**Reply by Kajalagg98** [reply]
Okay, cool.
